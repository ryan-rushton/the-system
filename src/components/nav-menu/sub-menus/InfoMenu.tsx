import clsx from 'clsx';
import { useCallback, useContext } from 'react';
import { useTranslation } from 'react-i18next';
import { AppContext, SystemContext, systemSize } from '../../../context/SystemContext';
import styles from './InfoMenu.module.scss';

/**
 * An info menu for the system. This shows a bunch of stats like km per pixel for different distances
 * (not all distances are equal by default). It also has buttons to show the red orbit lines and to
 * normalise the distances per pixel.
 */
export function InfoMenu({
  isVisible,
  orbitsVisible,
  onChangeSystemSize,
  onOrbitsVisibleChange,
}: {
  /** Whether the component is currently visible. */
  isVisible: boolean;
  /** Whether the red orbit lines are visible */
  orbitsVisible: boolean;
  /** Change handler for showing the red orbit lines. */
  onOrbitsVisibleChange(newOrbitsVisible: boolean): void;
  /** Change handler for normalising km per pixel. */
  onChangeSystemSize(newContext: SystemContext): void;
}) {
  const context = useContext(AppContext);
  const { t } = useTranslation();
  const { orbitalPeriodMultiplier, distanceMultiplier, sizeMultiplier, satelliteDist } = context.multipliers;

  const secondsInOneDay = 1 * orbitalPeriodMultiplier;
  const kmPerPixelDistance = Math.round(1 / distanceMultiplier).toLocaleString();
  const kmPerPixelSatellite = Math.round(1 / satelliteDist).toLocaleString();
  const kmPerPixelSize = Math.round(1 / sizeMultiplier).toLocaleString();

  const onOrbitChangeClick = useCallback(
    () => onOrbitsVisibleChange(!orbitsVisible),
    [onOrbitsVisibleChange, orbitsVisible],
  );

  const onSizeChangeClick = useCallback(() => {
    const { enhancedVisibility, evenSpace } = systemSize;
    const systemSizeContext = evenSpace === context ? enhancedVisibility : evenSpace;
    onChangeSystemSize(systemSizeContext);
  }, [context, onChangeSystemSize]);

  return (
    <div data-testid="info-menu">
      <div className={styles.heading}>{t('infoMenu.time')}</div>
      <div data-testid="time" className={styles.stat}>
        {t('infoMenu.secondsForOneDay', { seconds: secondsInOneDay })}
      </div>
      <div className={styles.heading}>{t('infoMenu.distanceBetweenPlanets')}</div>
      <div data-testid="planet-distance" className={styles.stat}>
        {t('infoMenu.kmPerPixel', { km: kmPerPixelDistance })}
      </div>
      <div className={styles.heading}>{t('infoMenu.distanceBetweenPlanetsAndMoons')}</div>
      <div data-testid="satellite-distance" className={styles.stat}>
        {t('infoMenu.kmPerPixel', { km: kmPerPixelSatellite })}
      </div>
      <div className={styles.heading}>{t('infoMenu.planetSize')}</div>
      <div data-testid="planet-size" className={styles.stat}>
        {t('infoMenu.kmPerPixel', { km: kmPerPixelSize })}
      </div>
      <div className={styles.buttonWrapper}>
        <button
          className={clsx(styles.button, { [styles.buttonActive]: systemSize.evenSpace === context })}
          onClick={onSizeChangeClick}
          tabIndex={isVisible ? 0 : undefined}
          aria-label={t('infoMenu.normaliseDistance') || undefined}
        >
          {t('infoMenu.normaliseDistance')}
        </button>
      </div>
      <div className={styles.buttonWrapper}>
        <button
          className={clsx(styles.button, { [styles.buttonActive]: orbitsVisible })}
          onClick={onOrbitChangeClick}
          tabIndex={isVisible ? 0 : undefined}
          aria-label={t('infoMenu.showOrbits') || undefined}
        >
          {t('infoMenu.showOrbits')}
        </button>
      </div>
    </div>
  );
}
