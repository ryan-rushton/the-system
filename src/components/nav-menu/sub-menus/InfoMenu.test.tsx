import { cleanup, fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import AppContext, { systemSize } from '../../../context/SystemContext';
import InfoMenu from './InfoMenu';

describe('InfoMenu', () => {
  afterEach(cleanup);

  test('it renders', () => {
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />);
    const element = screen.getByTestId('info-menu');

    expect(element).toBeInTheDocument();
  });

  test('it matches snapshot', () => {
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />);

    expect(screen.getByTestId('info-menu')).toMatchSnapshot();
  });

  test('it changes orbits visible when Show Orbits clicked', () => {
    const onOrbitsChange = jest.fn();
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={onOrbitsChange} onChangeSystemSize={jest.fn()} />);
    userEvent.click(screen.getByText(/^Show Orbits$/));

    expect(onOrbitsChange).toBeCalledTimes(1);
    expect(onOrbitsChange).toBeCalledWith(true);
  });

  test('it changes orbits visible when Show Orbits enter pressed', () => {
    const onOrbitsChange = jest.fn();
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={onOrbitsChange} onChangeSystemSize={jest.fn()} />);

    fireEvent.keyDown(screen.getByText(/^Show Orbits$/), { key: 'Enter' });

    expect(onOrbitsChange).toBeCalledTimes(1);
    expect(onOrbitsChange).toBeCalledWith(true);
  });

  test('it changes system size when Normalise Distance clicked', () => {
    const onChangeSystemSize = jest.fn();
    render(
      <InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={onChangeSystemSize} />
    );
    userEvent.click(screen.getByText(/^Normalise Distance$/));

    expect(onChangeSystemSize).toBeCalledTimes(1);
    expect(onChangeSystemSize).toBeCalledWith(systemSize.evenSpace);
  });

  test('it changes system size when Normalise Distance enter pressed', () => {
    const onChangeSystemSize = jest.fn();
    render(
      <InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={onChangeSystemSize} />
    );
    fireEvent.keyDown(screen.getByText(/^Normalise Distance$/), { key: 'Enter' });

    expect(onChangeSystemSize).toBeCalledTimes(1);
    expect(onChangeSystemSize).toBeCalledWith(systemSize.evenSpace);
  });

  test('it has correct days per second for enhanced visibility context', () => {
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />);

    const timeDetails = screen.getByTestId('time');

    expect(timeDetails).toHaveTextContent('1 s = 1 day');
  });

  test('it has correct days per second for even size context', () => {
    render(
      <AppContext.Provider value={systemSize.evenSpace}>
        <InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />
      </AppContext.Provider>
    );

    const timeDetails = screen.getByTestId('time');

    expect(timeDetails).toHaveTextContent('20 s = 1 day');
  });

  test('it has correct planet distance for enhanced visibility context', () => {
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />);

    const timeDetails = screen.getByTestId('planet-distance');

    expect(timeDetails).toHaveTextContent('1 pixel = 100,000 km');
  });

  test('it has correct planet distance for even size context', () => {
    render(
      <AppContext.Provider value={systemSize.evenSpace}>
        <InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />
      </AppContext.Provider>
    );

    const timeDetails = screen.getByTestId('planet-distance');

    expect(timeDetails).toHaveTextContent('1 pixel = 2,000 km');
  });

  test('it has correct satellite distance for enhanced visibility context', () => {
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />);

    const timeDetails = screen.getByTestId('satellite-distance');

    expect(timeDetails).toHaveTextContent('1 pixel = 10,000 km');
  });

  test('it has correct satellite distance for even size context', () => {
    render(
      <AppContext.Provider value={systemSize.evenSpace}>
        <InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />
      </AppContext.Provider>
    );

    const timeDetails = screen.getByTestId('satellite-distance');

    expect(timeDetails).toHaveTextContent('1 pixel = 2,000 km');
  });

  test('it has correct planet size for enhanced visibility context', () => {
    render(<InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />);

    const timeDetails = screen.getByTestId('planet-size');

    expect(timeDetails).toHaveTextContent('1 pixel = 2,000 km');
  });

  test('it has correct planet size for even size context', () => {
    render(
      <AppContext.Provider value={systemSize.evenSpace}>
        <InfoMenu orbitsVisible={false} onOrbitsVisibleChange={jest.fn()} onChangeSystemSize={jest.fn()} />
      </AppContext.Provider>
    );

    const timeDetails = screen.getByTestId('planet-size');

    expect(timeDetails).toHaveTextContent('1 pixel = 2,000 km');
  });
});
