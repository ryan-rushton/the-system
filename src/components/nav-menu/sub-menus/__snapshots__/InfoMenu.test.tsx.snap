// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoMenu it matches snapshot 1`] = `
<div
  data-testid="info-menu"
>
  <div
    class="heading"
  >
    Time
  </div>
  <div
    class="stat"
    data-testid="time"
  >
    1 s = 1 day
  </div>
  <div
    class="heading"
  >
    Distance Between Planets
  </div>
  <div
    class="stat"
    data-testid="planet-distance"
  >
    1 pixel = 100,000 km
  </div>
  <div
    class="heading"
  >
    Distance Between Planets And Moons
  </div>
  <div
    class="stat"
    data-testid="satellite-distance"
  >
    1 pixel = 10,000 km
  </div>
  <div
    class="heading"
  >
    Planet Size
  </div>
  <div
    class="stat"
    data-testid="planet-size"
  >
    1 pixel = 2,000 km
  </div>
  <div
    class="buttonWrapper"
  >
    <div
      aria-label="Normalise Distance"
      class="button"
      role="button"
      tabindex="0"
    >
      Normalise Distance
    </div>
  </div>
  <div
    class="buttonWrapper"
  >
    <div
      aria-label="Show Orbits"
      class="button"
      role="button"
      tabindex="0"
    >
      Show Orbits
    </div>
  </div>
</div>
`;
